ledger/catchpointwriter.go

```diff
@@ -138,93 +152,110 @@ func (cw *catchpointWriter) Abort() error {
 	return os.Remove(cw.filePath)
 }
 
+// WriteStep works for a short period of time (determined by stepCtx) to get
+// some more data (accounts/resources/kvpairs) by using readDatabaseStep, and
+// write that data to the open tar file in cw.tar.  The writing is done in
+// asyncWriter, so that it can proceed concurrently with reading the data from
+// the db. asyncWriter only runs long enough to process the data read during a
+// single call to WriteStep, and WriteStep ensures that asyncWriter has finished
+// writing by waiting for it in a defer block, collecting any errors that may
+// have occurred during writing.  Therefore, WriteStep looks like a simple
+// synchronous function to its callers.
 func (cw *catchpointWriter) WriteStep(stepCtx context.Context) (more bool, err error) {
 	// have we timed-out / canceled by that point ?
 	if more, err = hasContextDeadlineExceeded(stepCtx); more || err != nil {
 		return
 	}
 
-	writerRequest := make(chan catchpointFileBalancesChunkV6, 1)
+	writerRequest := make(chan catchpointFileChunkV6, 1)
 	writerResponse := make(chan error, 2)
-	go cw.asyncWriter(writerRequest, writerResponse, cw.balancesChunkNum, cw.numAccountsProcessed)
+	go cw.asyncWriter(writerRequest, writerResponse, cw.chunkNum)
 	defer func() {
+		// For simplicity, all cleanup is done once, here. The writerRequest is
+		// closed, signaling asyncWriter that it can exit, and then
+		// writerResponse is drained, ensuring any problems from asyncWriter are
+		// noted (and that the writing is done).
 		close(writerRequest)
-		// wait for the writerResponse to close.
+	DRAIN:
```

Comment:
It's certainly a bit of a refactor.  We used to try to determine ahead of time an exit condition for the loop, at first it was number of chunks, then it was number of accounts, and I would have been trying to do something that checked accounts _and_ kvstore. I refactored so that the iterators return empty values when they complete.  That made handling them easier.  But it also means that cleanup is easier to do outside (closing the tar file when readDatabaseStep stops returning results).  There was also duplicated work where we would check for errors on the "happy path" _and_ in the defer block to handle the "sad paths".  I changed it handle both cases in the defer.  Which meant that path had to be willing to read/drain an arbitrary number of errors. (Well, I think it can only be two, because of the bounds on channel buffers, but I thought this was the future-proof way to write it.)